server/pbenchinacan/deploy

```diff
@@ -77,44 +77,73 @@ fi
+        --volume "${SRV_PBENCH}":/srv/pbench:Z \
         --entrypoint ln \
         ${PB_SERVER_IMAGE} \
-        -sf ${NGINX_FAVICON} /srv/pbench/public_html/favicon.ico
+        -sf "${NGINX_FAVICON}" /srv/pbench/public_html/favicon.ico
 else
     echo "Warning:  deploying the Pbench Server without a Dashboard" >&2
     podman run \
         --rm \
-        --volume ${SRV_PBENCH}:/srv/pbench:Z \
+        --volume "${SRV_PBENCH}":/srv/pbench:Z \
         --entrypoint mkdir \
         ${PB_SERVER_IMAGE} \
         -m 0755 -p /srv/pbench/public_html/dashboard
 fi
 
+# Create a systemd service file for the Pbench Server
+# (This was originally generated by Podman and then heavily edited.)
+cat >"${PB_DEPLOY_FILES}/pbench-server.service" <<- EOD
+	[Unit]
+	Description=Containerized Pbench Server
+	Documentation=https://distributed-system-analysis.github.io/pbench/
+	Documentation=https://pbench.readthedocs.io/en/latest/Server/API/README.html
+	Documentation=https://github.com/distributed-system-analysis/pbench
+	Wants=network-online.target
+	After=network-online.target
+	RequiresMountsFor=%t/containers
+
+	[Service]
+	Environment=PODMAN_SYSTEMD_UNIT=%n
+	Restart=on-failure
+	TimeoutStopSec=30
+	ExecStartPre=/bin/rm -f %t/%n.ctr-id
+	ExecStart=/usr/bin/podman run \
+	  --cidfile=%t/%n.ctr-id \
+	  --cgroups=no-conmon \
+	  --sdnotify=conmon \
+	  --replace \
+	  --detach \
+	  --name "${PB_SERVER_CONTAINER_NAME}" \
+	  --network host \
+	  --rm \
+	  --volume "${PB_DEPLOY_FILES}"/etc/rsyslog.conf:/etc/rsyslog.conf:Z \
+	  --volume "${PB_DEPLOY_FILES}"/etc/rsyslog.d:/etc/rsyslog.d:Z \
+	  --volume "${PB_SSL_CERT_FILE}":/etc/pki/tls/certs/pbench-server.crt:Z \
+	  --volume "${PB_SSL_KEY_FILE}":/etc/pki/tls/private/pbench-server.key:Z \
+	  --volume "${PB_DEPLOY_FILES}"/pbench-server.cfg:/opt/pbench-server/lib/config/pbench-server.cfg:Z \
+	  --volume "${SRV_PBENCH}":/srv/pbench:Z \
+	  ${PB_SERVER_PODMAN_SWITCHES} \
+	  ${PB_SERVER_IMAGE}
+	ExecStop=/usr/bin/podman stop --ignore --cidfile=%t/%n.ctr-id
+	ExecStopPost=/usr/bin/podman rm -f --ignore --cidfile=%t/%n.ctr-id
+	Type=notify
+	NotifyAccess=all
+
+	[Install]
+	WantedBy=default.target
+EOD
+
 # Run the Pbench Server.
-#
-# See NOTE above on the first Podman run invocation regarding pull policy.
-podman run \
-    --detach \
-    --name ${PB_SERVER_CONTAINER_NAME} \
-    --network host \
-    --rm \
-    --volume ${PB_DEPLOY_FILES}/etc/rsyslog.conf:/etc/rsyslog.conf:Z \
-    --volume ${PB_DEPLOY_FILES}/etc/rsyslog.d:/etc/rsyslog.d:Z \
-    --volume ${PB_SSL_CERT_FILE}:/etc/pki/tls/certs/pbench-server.crt:Z \
-    --volume ${PB_SSL_KEY_FILE}:/etc/pki/tls/private/pbench-server.key:Z \
-    --volume ${PB_DEPLOY_FILES}/pbench-server.cfg:/opt/pbench-server/lib/config/pbench-server.cfg:Z \
-    --volume ${SRV_PBENCH}:/srv/pbench:Z \
-    ${PB_SERVER_PODMAN_SWITCHES} \
-    ${PB_SERVER_IMAGE}
+systemctl --user enable --force --now "${PB_DEPLOY_FILES}/pbench-server.service"
```

Comment:
If the idea is to use this `deploy` file for the staging server... I wonder if

1. We might want to hack the service file, suggesting that `generate_service` might be offloaded into a separate script (or at least a function) so we can repeatedly deploy without overwriting the service file.
2. If we're *not* changing the service file, we don't need to regenerate it on each startup: offloading that into a separate script would let us do it only when necessary.

Reply:
My expectation is that the `deploy` script would only be used when creating _new_ deployments.  Updating existing deployments would be done with a `systemctl restart`, possibly after doing a `podman pull`.  However, establishing a new Staging server or a new Production server should be "as simple as" running this script.

So, toward that end, we need to create/generate the service file every time we run this script.  (That's true for new Staging/Production deployments as well as func-test ones; arguably, for development, it's somewhat less true, but in that case I don't think re-generating it every time really hurts.)

That said, my lack of faith extends to using Podman (or `systemd`) to (re)generate service files from existing containers.  Doing so as a starting point and then editing the service file is great (although, apparently, Podman has deprecated this feature...), but I just don't trust it as a general solution.  So, I would rather have the service file produced from a template, at which point I don't think it really helps to separate the template from the script which expands it.

All that said, there is one thing which the `deploy` script does which isn't handled by a simple `podman pull` and `systemctl restart`, and that's the deployment of a new Dashboard, which, sadly, is slightly less than trivial.

So, perhaps what we need is a `deploy` script wrapped around a `redeploy` script, where the former does all the "one-time" stuff (like generating the service file) and the latter does the "every-time" stuff (like deploying the Dashboard and starting the service)?

Reply:
> So, perhaps what we need is a `deploy` script wrapped around a `redeploy` script, where the former does all the "one-time" stuff (like generating the service file) and the latter does the "every-time" stuff (like deploying the Dashboard and starting the service)?

Yeah, I was thinking about the dashboard, too; and that model would probably work.